### PR TITLE
Cortex-M MPU driver and pseudo-module for guarding against stack overflows

### PR DESCRIPTION
**Pull request by hexluthor**
A 32-byte region at the bottom of the running stack is made read-only by the MPU. If a stack overflow occurs, a memory management exception is triggered before any data below the stack is overwritten. Depending on the alignment of the individual stacks, this comes at a cost of 32-63 bytes per thread. For this reason, the feature is an optional pseudo-module and is not enabled by default.

As a future PR, I'd like to define an attribute for all statically-allocated stacks that instructs the linker to align them to a 32-byte boundary if this module is used, and would otherwise be just a placeholder or "no-op".

### CODE REVIEW

**Comment by kaspar030** [comment]
nice!

File: Makefile.pseudomodules [hunk 1]
```diff
@@ -33,6 +33,7 @@ PSEUDOMODULES += lwip_stats
 PSEUDOMODULES += lwip_tcp
 PSEUDOMODULES += lwip_udp
 PSEUDOMODULES += lwip_udplite
+PSEUDOMODULES += mpu_stack_guard
```
Comment (kaspar030) [inline Makefile.pseudomodules]: cortexm_ ?

**Comment by gebart** [comment]
Cool! We need a test though, do you have an application for testing? Otherwise we can write a recursive function which will run out of stack and put it under tests/stack_overflow or something. 
The same test would be interesting for the isr stack since the cortexm isr stack relocation pr was merged.

File: cpu/cortexm_common/mpu.c [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+}
+
+int mpu_enable(void) {
+#if __MPU_PRESENT
+    MPU->CTRL |= MPU_CTRL_PRIVDEFENA_Msk | MPU_CTRL_ENABLE_Msk;
+
+    /* Enable the memory fault exception */
+    SCB->SHCSR |= SCB_SHCSR_MEMFAULTENA_Msk;
+
+    return 0;
```
Comment (kaspar030) [inline cpu/cortexm_common/mpu.c]: This returns 0 on success, 1 on "not available", right?
I find the opposite more intuitive.

File: cpu/cortexm_common/mpu.c [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+    MPU->RNR  = region;
+    MPU->RBAR = base & MPU_RBAR_ADDR_Msk;
+    MPU->RASR = attr | MPU_RASR_ENABLE_Msk;
+
+    if (!mpu_enabled()) {
+        mpu_enable();
+    }
+
+    return 0;
```
Comment (kaspar030) [inline cpu/cortexm_common/mpu.c]: same.

File: Makefile.pseudomodules (hunk 1 above)
Comment (hexluthor) [inline Makefile.pseudomodules]: I considered that, but wanted to leave room for other MPUs to be used in the same way.

**Comment by hexluthor** [comment]
@gebart I had a feeling someone would ask for a test app. I don't have one yet but will think about it.

**Comment by gebart** [comment]
@kaspar030 @hexluthor 
wouldn't it be more consistent to use negative return values for errors, like in the periph drivers do it?

**Comment by gebart** [comment]
Suggestion for a follow up PR: Mark the running stack non-executable.

**Comment by hexluthor** [comment]
@gebart RIOT seems pretty inconsistent about this but what you describe is also the Linux coding style so I'm going with it.

**Comment by hexluthor** [comment]
Here's a test app too. Tested on `cc2538dk`.

**Comment by kYc0o** [comment]
Great! tested on arduino-due and it works like a charm, so ACK. Please squash and set the flag for Murdock if possible.

**Comment by OlegHahm** [comment]
> @gebart RIOT seems pretty inconsistent about this 

Can you elaborate on where you found these inconsistencies?

**Comment by kYc0o** [comment]
We can wait for the next release on this one...

**Comment by hexluthor** [comment]
@OlegHahm I did something like `ack-grep 'return.*(success|fail)'`. In most cases functions return 0 on success, but there are some that return 1. Didn't really investigate beyond the `grep`.

**Comment by OlegHahm** [comment]
I would guess that in some cases (maybe not all) this is okay, according to https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions#return-values

**Comment by miri64** [comment]
Is there anything to discuss in this PR (it was ACKed in https://github.com/RIOT-OS/RIOT/pull/5564#issuecomment-234536283)

File: core/sched.c [hunk 4]
```diff
@@ -115,6 +119,16 @@ int sched_run(void)
     sched_active_pid = next_thread->pid;
     sched_active_thread = (volatile thread_t *) next_thread;
 
+#ifdef MODULE_MPU_STACK_GUARD
+    mpu_configure(
+        1,                                                /* MPU region 1 */
+        (uintptr_t)sched_active_thread->stack_start + 31, /* Base Address (rounded up) */
+        MPU_ATTR(1, AP_RO_RO, 0, 1, 0, 1, MPU_SIZE_32B)   /* Attributes and Size */
```
Comment (kaspar030) [inline core/sched.c]: This prevents overwriting the lowest byte of the stack?

Comment (kaspar030) [inline core/sched.c]: please ignore the noise, just read the PR description again. ;)

**Comment by miri64** [comment]
So can we merge?

**Comment by hexluthor** [comment]
I see no reason to not merge it (but I'm also the author).

File: cpu/cortexm_common/include/mpu.h [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+ * @param[in]   ap      Access Permission word, e.g. AP_RO_RO
+ * @param[in]   tex     Type Extension Field
+ * @param[in]   c       Cacheable bit
+ * @param[in]   b       Bufferable bit
+ * @param[in]   s       Sub-Region Disable (SRD) field
+ * @param[in]   size    region size code, e.g. MPU_SIZE_32B
+ *
+ * @return combined region attribute word
+ */
+#define MPU_ATTR(xn, ap, tex, c, b, s, size) ( \
```
Comment (kaspar030) [inline cpu/cortexm_common/include/mpu.h]: Conventions say multiline macros should be static inline functions. Would that optimize to the same code in this case?

Comment (hexluthor) [inline cpu/cortexm_common/include/mpu.h]: I prefer the macro but yield to convention. They are binary equivalent for `tests/mpu_stack_guard`.

**Comment by miri64** [comment]
@kYc0o

> Rebase if possible and maybe we should let Murdock look into it again, because this PR is kind of old and several things were merged in the meantime.

Murdock seemed to be happy last time I tried it and there are no conflicts so why rebase?

**Comment by kYc0o** [comment]
Oh, I didn't know you tried afterwards. It should be OK then. Yeah for the conflicts I agree too. So just this "inconsistency" about the return on failure but the arguments are "sometimes 0 is valid, sometimes not", so that's maybe another discussion. I propose to merge this as is and when an agreement about what are the "right" return errors we can change them all, including this one.

**Comment by miri64** [comment]
> Oh, I didn't know you tried afterwards. 

I'm not sure this is clear now: I did not test this PR, I just observed Murdock's results ;-).

**Comment by kYc0o** [comment]
Yes, it was me who tested successfully ;)

**Comment by hexluthor** [comment]
Ok I
 * Rebased onto `master`.
 * Whitelisted new Cortex-M boards `arduino-duemilanove`, `arduino-uno`, `arduino-zero`, `nucleo-f207`, `nucleo-f446`, `sodaq-autonomo` in `tests/mpu_stack_guard`.
 * Corrected a doxygen typo about return values. Besides `mpu_enabled()`, all functions return 0 on success, -1 on failure.

**Comment by kYc0o** [comment]
It seems OK for me now, I'd re-ACK but I'd like the second commit divided into several for the files you're changing, because now it changes several files and the commit message it's not prefixed.

**Comment by hexluthor** [comment]
@kYc0o ok I committed the test app separately. Not sure what else could be divided.

**Comment by kYc0o** [comment]
Maybe the commits for `core` and `cpu`... I can also see some commit on another test but I don't know if split it too... Since this PR is touching "delicate" files I'd like to have them in a separated commit.

**Comment by kYc0o** [comment]
and go!

**Pull request merged**